src/ffi/ios.rs

```diff
@@ -1,40 +1,130 @@
+        .context("parse url")?;
+    let host = host_url.host_str()
+        .context("no host in request")?;
+
+    let mut req = Request::builder()
+        .method(Method::from_bytes(request.method.as_bytes()).context("invalid method")?)
+        .header("Host", host)
         .version(http::Version::HTTP_10)
-        .header("Host", domain)
-        .body(vec![])
-        .expect("construct request");
-
-    match client.send(req) {
-        Ok(s) => to_cf_str(format!(
-            "Result is: {:?}",
-            s.map(|raw| String::from_utf8(raw).expect("decode response as utf8"))
-        )),
-        Err(e) => to_cf_str(format!("Error while getting result: {}", e)),
+        .uri(&request.url)
+        .body(request.body)
+        .context("invalid request")?;
+
+    let hm = req.headers_mut();
+    for (key, values) in request.headers {
+        for value in values {
+            hm.append(&HeaderName::from_bytes(&key.as_bytes())?,
+                      HeaderValue::from_bytes(value.as_bytes())?);
+        }
+    }
+
+    info!("Parsed request is: {:?}", req);
+
+    Client::new(request.dict_dir.into()).send(req).try_into()
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ReturnStruct {
+    error: Option<JSONError>,
+    response: Option<Response>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Response {
+    status: u16,
+    headers: HashMap<String, Vec<String>>,
+    body: Vec<u8>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct JSONError {
+    error_string: String,
+    error_context: Option<String>,
+}
+
+impl TryFrom<Result<http::response::Response<Vec<u8>>>> for ReturnStruct {
+    type Error = anyhow::Error;
+
+    fn try_from(resp_result: Result<http::response::Response<Vec<u8>>>) -> Result<Self, anyhow::Error> {
+        match resp_result {
+            Ok(resp) => {
+                let mut headers = HashMap::new();
+                for (k, v) in resp.headers() {
+                    headers.insert(k.to_string(), vec![v.to_str()?.to_string()]);
+                }
+                Ok(Self {
+                    error: None,
+                    response: Some(Response {
+                        status: resp.status().as_u16(),
+                        headers,
+                        body: resp.body().clone(),
+                    }),
+                })
+            }
+            Err(e) => Ok(ReturnStruct {
+                error: Some(JSONError {
+                    error_string: e.to_string(),
+                    error_context: None,
+                }),
+                response: None,
+            }),
+        }
+    }
+}
+
+impl ReturnStruct {
+    pub fn err(error: String) -> ReturnStruct {
+        ReturnStruct {
+            error: Some(JSONError {
+                error_string: error.to_string(),
+                error_context: None,
+            }),
+            response: None,
+        }
+    }
+
+    pub fn to_json(&self) -> String {
```

Comment:
You can also inline this single call, thus separating concerns of serialization from the ReturnStruct's implementation.

And yeah, I'm all for reducing the number of exposed functions, the less there is, the safer it is IMO :)